sharding/mainchain/smc_client.go

```diff
@@ -0,0 +1,190 @@
+// Package mainchain defines services that interacts with a Geth node via RPC.
+// This package is useful for an actor in a sharded system to interact with
+// a Sharding Manager Contract.
+package mainchain
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"sync"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/sharding/contracts"
+)
+
+// ClientIdentifier tells us what client the node we interact with over RPC is running.
+const ClientIdentifier = "geth"
+
+// Client contains useful methods for a sharding node to interact with
+// an Ethereum client running on the mainchain.
+type Client interface {
+	Account() *accounts.Account
+	CreateTXOpts(value *big.Int) (*bind.TransactOpts, error)
+	SMCCaller() *contracts.SMCCaller
+	SMCTransactor() *contracts.SMCTransactor
+	ChainReader() ethereum.ChainReader
+	DepositFlag() bool
+	SetDepositFlag(deposit bool)
```

Comment:
This should not be changed after initialization. Setting the flag again was only set in our tests, but we should have just created another client or some other method to test the functionality with a different deposit flag value.